File: frontend/src/services/terminal.ts
/**
 * Terminal service for managing terminal sessions
 */

import api from './api';

export interface TerminalSession {
  session_id: string;
  host_id: string;
  user_id: string;
  state: string;
  rows: number;
  cols: number;
  created_at: string;
  last_activity: string;
  connected: boolean;
}

export interface TerminalMessage {
  type: string;
  session_id: string;
  data: string;
  timestamp?: string;
}

/**
 * Terminal service
 */
export const terminalService = {
  /**
   * Create a new terminal session
   */
  async createSession(
    hostId: string,
    rows: number = 24,
    cols: number = 80,
    env?: Record<string, string>
  ): Promise<TerminalSession> {
    try {
      const response = await api.post<TerminalSession>(`/terminal/sessions/${hostId}`, {
        rows,
        cols,
        env,
      });
      return response.data;
    } catch (error) {
      throw new Error(`Failed to create terminal session: ${error}`);
    }
  },

  /**
   * List all active terminal sessions
   */
  async listSessions(): Promise<TerminalSession[]> {
    try {
      const response = await api.get<{ items: TerminalSession[]; total: number }>(
        '/terminal/sessions'
      );
      return response.data.items;
    } catch (error) {
      throw new Error(`Failed to list terminal sessions: ${error}`);
    }
  },

  /**
   * Get a specific terminal session
   */
  async getSession(sessionId: string): Promise<TerminalSession> {
    try {
      const response = await api.get<TerminalSession>(`/terminal/sessions/${sessionId}`);
      return response.data;
    } catch (error) {
      throw new Error(`Failed to get terminal session: ${error}`);
    }
  },

  /**
   * Delete/close a terminal session
   */
  async deleteSession(sessionId: string): Promise<void> {
    try {
      await api.delete(`/terminal/sessions/${sessionId}`);
    } catch (error) {
      throw new Error(`Failed to delete terminal session: ${error}`);
    }
  },

  /**
   * Resize terminal
   */
  async resizeTerminal(
    sessionId: string,
    rows: number,
    cols: number
  ): Promise<void> {
    try {
      await api.post(`/terminal/sessions/${sessionId}/resize`, {
        rows,
        cols,
      });
    } catch (error) {
      throw new Error(`Failed to resize terminal: ${error}`);
    }
  },

  /**
   * Execute a command in terminal
   */
  async executeCommand(
    sessionId: string,
    command: string,
    timeout?: number
  ): Promise<void> {
    try {
      await api.post(`/terminal/sessions/${sessionId}/execute`, {
        command,
        timeout,
      });
    } catch (error) {
      throw new Error(`Failed to execute command: ${error}`);
    }
  },

  /**
   * Get terminal preferences
   */
  async getPreferences(): Promise<any> {
    try {
      const preferences = localStorage.getItem('termone_terminal_preferences');
      return preferences ? JSON.parse(preferences) : getDefaultPreferences();
    } catch (error) {
      return getDefaultPreferences();
    }
  },

  /**
   * Save terminal preferences
   */
  async savePreferences(preferences: any): Promise<void> {
    try {
      localStorage.setItem('termone_terminal_preferences', JSON.stringify(preferences));
    } catch (error) {
      throw new Error(`Failed to save terminal preferences: ${error}`);
    }
  },

  /**
   * Get theme
   */
  async getTheme(): Promise<string> {
    try {
      const theme = localStorage.getItem('termone_terminal_theme');
      return theme || 'dracula';
    } catch (error) {
      return 'dracula';
    }
  },

  /**
   * Set theme
   */
  async setTheme(theme: string): Promise<void> {
    try {
      localStorage.setItem('termone_terminal_theme', theme);
    } catch (error) {
      throw new Error(`Failed to set terminal theme: ${error}`);
    }
  },

  /**
   * Get font size
   */
  async getFontSize(): Promise<number> {
    try {
      const size = localStorage.getItem('termone_terminal_font_size');
      return size ? parseInt(size) : 14;
    } catch (error) {
      return 14;
    }
  },

  /**
   * Set font size
   */
  async setFontSize(size: number): Promise<void> {
    try {
      localStorage.setItem('termone_terminal_font_size', size.toString());
    } catch (error) {
      throw new Error(`Failed to set font size: ${error}`);
    }
  },

  /**
   * Get font family
   */
  async getFontFamily(): Promise<string> {
    try {
      const family = localStorage.getItem('termone_terminal_font_family');
      return family || "'Courier New', monospace";
    } catch (error) {
      return "'Courier New', monospace";
    }
  },

  /**
   * Set font family
   */
  async setFontFamily(family: string): Promise<void> {
    try {
      localStorage.setItem('termone_terminal_font_family', family);
    } catch (error) {
      throw new Error(`Failed to set font family: ${error}`);
    }
  },
};

function getDefaultPreferences() {
  return {
    theme: 'dracula',
    fontSize: 14,
    fontFamily: "'Courier New', monospace",
    bellEnabled: true,
    bellType: 'visual',
  };
}

export default terminalService;

File: backend/src/schemas/terminal.py
"""Pydantic schemas for terminal API."""

from typing import Optional, List
from datetime import datetime
from enum import Enum

from pydantic import BaseModel, Field


class TerminalMessageType(str, Enum):
    """Terminal message types."""

    OUTPUT = "output"
    INPUT = "input"
    CONTROL = "control"
    ERROR = "error"
    PING = "ping"
    PONG = "pong"


class TerminalSessionState(str, Enum):
    """Terminal session states."""

    CREATED = "created"
    CONNECTED = "connected"
    RUNNING = "running"
    DISCONNECTED = "disconnected"
    CLOSED = "closed"
    ERROR = "error"


class TerminalSessionCreate(BaseModel):
    """Create terminal session request."""

    rows: int = Field(24, ge=1, le=500, description="Terminal rows")
    cols: int = Field(80, ge=1, le=500, description="Terminal columns")
    env: Optional[dict] = Field(None, description="Environment variables")


class TerminalSessionResponse(BaseModel):
    """Terminal session response."""

    session_id: str = Field(..., description="Session ID")
    host_id: str = Field(..., description="Host ID")
    user_id: str = Field(..., description="User ID")
    state: TerminalSessionState = Field(..., description="Session state")
    rows: int = Field(..., description="Terminal rows")
    cols: int = Field(..., description="Terminal columns")
    created_at: datetime = Field(..., description="Creation timestamp")
    last_activity: datetime = Field(..., description="Last activity timestamp")
    connected: bool = Field(..., description="Is session connected")

    class Config:
        from_attributes = True


class TerminalSessionListResponse(BaseModel):
    """Terminal sessions list response."""

    items: List[TerminalSessionResponse] = Field(..., description="List of sessions")
    total: int = Field(..., description="Total number of sessions")


class TerminalMessage(BaseModel):
    """Terminal message."""

    type: TerminalMessageType = Field(..., description="Message type")
    session_id: str = Field(..., description="Session ID")
    data: str = Field(..., description="Message data")
    timestamp: Optional[datetime] = Field(None, description="Message timestamp")


class TerminalResizeRequest(BaseModel):
    """Terminal resize request."""

    rows: int = Field(..., ge=1, le=500, description="New terminal rows")
    cols: int = Field(..., ge=1, le=500, description="New terminal columns")


class TerminalInputRequest(BaseModel):
    """Terminal input request."""

    data: str = Field(..., description="Input data to send to terminal")


class TerminalExecuteRequest(BaseModel):
    """Terminal command execution request."""

    command: str = Field(..., description="Command to execute")
    timeout: Optional[int] = Field(None, ge=1, le=300, description="Command timeout in seconds")


class TerminalOutputData(BaseModel):
    """Terminal output data."""

    data: str = Field(..., description="Output data")
    timestamp: datetime = Field(..., description="Output timestamp")
    size: int = Field(..., description="Data size in bytes")


class TerminalReconnectRequest(BaseModel):
    """Terminal reconnection request."""

    session_id: str = Field(..., description="Session ID to reconnect")


class TerminalTheme(str, Enum):
    """Terminal themes."""

    DRACULA = "dracula"
    SOLARIZED_DARK = "solarized-dark"
    SOLARIZED_LIGHT = "solarized-light"
    NORD = "nord"
    GRUVBOX = "gruvbox"
    ONE_DARK = "one-dark"
    MONOKAI = "monokai"


class TerminalPreferences(BaseModel):
    """Terminal preferences."""

    theme: TerminalTheme = Field(TerminalTheme.DRACULA, description="Terminal theme")
    font_size: int = Field(14, ge=8, le=32, description="Font size in pixels")
    font_family: str = Field("'Courier New', monospace", description="Font family")
    bell_enabled: bool = Field(True, description="Enable terminal bell")
    bell_type: str = Field("visual", description="Bell type: visual or audio")


class TerminalError(BaseModel):
    """Terminal error response."""

    code: str = Field(..., description="Error code")
    message: str = Field(..., description="Error message")
    timestamp: datetime = Field(..., description="Error timestamp")
